### PR TITLE
Rewrite

### PR DESCRIPTION
**Pull request by niahmiah**
This is pretty much a rewrite of the service to provide a robust base server to build upon.

- Ripped hapi out, as we are no longer pursuing it.

- Adds a request ID to each request, and adds it to the header. This is useful for tracing a request through multiple services.

- Logs each request ID, method, and originalUrl

- Logs profile info for each request

- Addresses the prior lack of asynchronous start and stop commands, which are necessary to allow starting of dependencies before listening for connections, as well as being useful for graceful shutdown hooks.

- Also added in Cnn-messaging by default, which is only loaded and started when passed a config.

- Added test framework and coverage tools. Ready for TravisCI.

- Removed things that didn't seem to belong here. Probably worth debating. Let's discuss.

I am not sure which service we have that already use this other than the cnn-starter-api, but I will gladly refactor them if this gets merged.

### CODE REVIEW

**Comment by adslaton** [comment]
Please change it so this PR to would be merged into develop not master.

**Comment by pborrego** [comment]
I believe the generated app code for cnn-starter-app uses cnn-server as well: https://github.com/cnnlabs/cnn-starter-app/blob/develop/src/structure/dependencies.js

**Comment by niahmiah** [comment]
Run `docker-compose up –d` to start rabbitmq for most recent tests…

BTW, travis CI runs the tests automatically on each commit now

Thanks,

Ian Patton
Application Architect
CNN Enabling Team

**Comment by adslaton** [comment]
@niahmiah 

I get this error:

```
$ docker-compose up –d
Traceback (most recent call last):
  File "docker-compose", line 3, in <module>
  File "compose/cli/main.py", line 68, in main
  File "compose/cli/main.py", line 118, in perform_command
  File "compose/cli/main.py", line 926, in up
  File "compose/project.py", line 398, in up
  File "compose/project.py", line 183, in get_services_without_duplicate
  File "compose/project.py", line 171, in get_services
  File "compose/project.py", line 142, in get_service
  File "compose/project.py", line 632, in __init__
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 0: ordinal not in range(128)
Failed to execute script docker-compose
```

My docker version:

```
$ docker -v
Docker version 17.06.2-ce, build cec0b72
```

**Comment by adslaton** [comment]
I was able to test with:

```
$ docker-compose up
Creating cnnserver_amqp_1 ...
Creating cnnserver_amqp_1 ... done
Attaching to cnnserver_amqp_1
```

There must be something wonky going on with my docker.

**Comment by adslaton** [comment]
Where are we with this PR?

File: src/index.js [hunk 1]
```diff
@@ -1,20 +1,117 @@
...
+      }
+      next();
+    });
+    this.app.disable('x-powered-by');
+    this.app.use(reqId);
+    this.app.use(reqLogger);
+    this.service = null;
+  }
+
+  async start() {
```
Comment (phapturner) [inline src/index.js]: node.js doesn't support async await fully until 7.10.1 [http://node.green/](node.green). can we put in a .nvmrc file with the supported node version?

**Pull request closed without merging**